include/boost/numeric/ublas/tensor/functions.hpp

```diff
@@ -24,814 +24,908 @@
-
-		if( detail::retrieve_extents( expr  ).empty() )
-			throw std::runtime_error("error in boost::numeric::ublas::conj: tensors should not be empty.");
-
-		auto a = tensor_type( expr );
-		auto c = tensor_complex_type( a.extents() );
-
-		std::transform( a.begin(), a.end(),  c.begin(), [](auto const& l){ return std::conj(l) ; }  );
-
-		return c;
-	}
-
-	/** @brief Extract the real component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class T, class D>
-	auto real(detail::tensor_expression<T,D> const& expr) {
-		return detail::make_unary_tensor_expression<T> (expr(), [] (auto const& l) { return std::real( l ); } );
-	}
-
-	/** @brief Extract the real component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class V, class E, class A, class F, class D>
-	auto real(detail::tensor_expression<tensor<std::complex<V>,E,F,A>,D> const& expr)
-	{
-		using tensor_complex_type = tensor<std::complex<V>,E,F,A>;
-		using tensor_type = tensor<V,E,F,typename storage_traits<A>::template rebind<V>>;
-
-		if( detail::retrieve_extents( expr  ).empty() )
-			throw std::runtime_error("error in boost::numeric::ublas::real: tensors should not be empty.");
-
-		auto a = tensor_complex_type( expr );
-		auto c = tensor_type( a.extents() );
-
-		std::transform( a.begin(), a.end(),  c.begin(), [](auto const& l){ return std::real(l) ; }  );
-
-		return c;
-	}
-
-
-	/** @brief Extract the imaginary component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class T, class D>
-	auto imag(detail::tensor_expression<T,D> const& lhs) {
-		return detail::make_unary_tensor_expression<T> (lhs(), [] (auto const& l) { return std::imag( l ); } );
-	}
-
-
-	/** @brief Extract the imag component of tensor elements within a tensor expression
-	 *
-	 * @param[in] lhs tensor expression
-	 * @returns   unary tensor expression
-	*/
-	template<class V, class E, class A, class F, class D>
-	auto imag(detail::tensor_expression<tensor<std::complex<V>,E,F,A>,D> const& expr)
-	{
-		using tensor_complex_type = tensor<std::complex<V>,E,F,A>;
-		using tensor_type = tensor<V,E,F,typename storage_traits<A>::template rebind<V>>;
-
-		if( detail::retrieve_extents( expr  ).empty() )
-			throw std::runtime_error("error in boost::numeric::ublas::real: tensors should not be empty.");
-
-		auto a = tensor_complex_type( expr );
-		auto c = tensor_type( a.extents() );
-
-		std::transform( a.begin(), a.end(),  c.begin(), [](auto const& l){ return std::imag(l) ; }  );
-
-		return c;
-	}
+    namespace detail{
+        // MSVC 14.1 bug
+        // cannot use ::_size inside constexpr function
+        // basic_extents does not contains _size member because of which
+        // it complains about it
```

Comment:
Can you use a `constexpr` function, sth. like 
```cpp
constexpr size_type size() const { return _size; }
```

Reply:
size is already `constexpr`. I think this is fixed in the compiler we are using.

Reply:
I am not quite sure what you mean.
You commented that due to a bug in MSVC 14.1, you cannot use the **public** member data `_size` directly. I am wondering why you do not make `_size` a **private** member data and use a constexpr function `size()`instead.

Reply:
```cpp
if constexpr( is_static_v<extents_type> ){
   if constexpr ( extents_type::_size ){....}
}else {
  ....
}
```
msvc 14.1 will say basic_extents<T> does not have `_size`. Which is not a proper behaviour of `constexpr`

Reply:
So you cannot define?
```cpp
constexpr size_type size() const { return _size; }
```
in `fixed_rank_extents` and use it later?

Reply:
it will then give error size() cannot be used at compile time. dynamic extents `size()` cannot be known at compile time. I think have used it before and the above error stopped me msvc 14.1 but I haven't checked for 14.2. That's why I took the alternative of defining functions.

Reply:
Hmmm. Okay. Thanks for clarifying.